### PR TITLE
[discovery] Add OpenStack cloud provider discovery data

### PR DESCRIPTION
**Pull request by artbin**
## Description

- Add the OpenAPI specification for OpenStack cloud provider discovery data.
- Add the `d8-cloud-provider-discovery-data` secret to `kube-system` namespace.
- Add a hook that is watching for `d8-cloud-provider-discovery-data` secret updates.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: discovery
type: feature
summary: Add OpenStack cloud provider discovery data.
impact_level: default
```

### CODE REVIEW

File: ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go [hunk 1]
```diff
@@ -84,3 +98,252 @@ func (d *Discoverer) InstanceTypes(_ context.Context) ([]v1alpha1.InstanceType,
 
 	return res, nil
 }
+
+func (d *Discoverer) DiscoveryData(ctx context.Context) ([]byte, error) {
```
Comment (zuzzas) [inline ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:102]: Consider implementing backoff logic.

https://pkg.go.dev/github.com/gophercloud/gophercloud

@name212

File: ee/modules/030-cloud-provider-openstack/hooks/handle_discovery_data_volume_types.go [hunk 2]
```diff
@@ -6,67 +6,37 @@ Licensed under the Deckhouse Platform Enterprise Edition (EE) license. See https
...
-		openstackVolumeTypes, err = getVolumeTypesArray()
-		if err != nil {
-			return err
-		}
-	}
+	"github.com/deckhouse/deckhouse/go_lib/cloud-data/apis/v1alpha1"
+)
 
-	storageClassesMap := make(map[string]string, len(openstackVolumeTypes))
+func handleDiscoveryDataVolumeTypes(input *go_hook.HookInput, volumeTypes []v1alpha1.OpenStackCloudProviderDiscoveryDataVolumeType) {
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/hooks/handle_discovery_data_volume_types.go:19]: We use this function in one place in discoverer hook. Let move this function in the hook file.

File: ee/modules/030-cloud-provider-openstack/hooks/handle_discovery_data_volume_types.go [hunk 3]
```diff
@@ -110,3 +78,8 @@ func getStorageClassName(value string) string {
 	// must start and end with an alphanumeric character
 	return strings.Trim(value, "-.")
 }
+
+type storageClass struct {
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/hooks/handle_discovery_data_volume_types.go:82]: with this struct and remove file

File: ee/modules/030-cloud-provider-openstack/hooks/init_openstack_envs.go [hunk 4]
```diff
@@ -7,38 +7,21 @@ package hooks
...
-	}
-
-	var volumeTypesList []string
-	for _, vt := range volumeTypes {
-		volumeTypesList = append(volumeTypesList, vt.Name)
-	}
-
-	return volumeTypesList, nil
-}
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/hooks/init_openstack_envs.go:16]: We do not need initialize openstack envs now. we've moved all operations with openstack from deckhouse to `data-discoverer` service. We can (or should) remove initialize function and hook.
Also, do not forget to  remove gophercloud library from deckhouse `go.mod`

File: ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go (hunk 1 above)
Comment (name212) [reply ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:102]: @artbin Let implement backoff logic for openstack request

File: ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go [hunk 5]
```diff
@@ -84,3 +98,252 @@ func (d *Discoverer) InstanceTypes(_ context.Context) ([]v1alpha1.InstanceType,
...
+	cctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	obj, err := d.k8sClient.Resource(schema.GroupVersionResource{Group: "", Version: "v1", Resource: "secrets"}).Namespace("kube-system").Get(cctx, "d8-provider-cluster-configuration", metav1.GetOptions{})
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:106]: We can use normal (not dynamic) client for secrets. Lets initialize regular go-client and pass to discoverer

File: ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go [hunk 6]
```diff
@@ -84,3 +98,252 @@ func (d *Discoverer) InstanceTypes(_ context.Context) ([]v1alpha1.InstanceType,
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to list security groups: %v", err)
+	}
+
+	groups, err := groups.ExtractGroups(allPages)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract security groups: %v", err)
+	}
+	if len(groups) == 0 {
+		return nil, errors.New("security groups list is empty")
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:297]: security groups list can be empty

File: ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go [hunk 7]
```diff
@@ -84,3 +98,252 @@ func (d *Discoverer) InstanceTypes(_ context.Context) ([]v1alpha1.InstanceType,
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to list images: %v", err)
+	}
+
+	images, err := images.ExtractImages(allPages)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract images: %v", err)
+	}
+	if len(images) == 0 {
+		return nil, errors.New("images list is empty")
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:329]: controversial, but images list can be empty

File: go_lib/cloud-data/reconciler.go [hunk 8]
```diff
@@ -274,3 +282,106 @@ func (c *Reconciler) cloudDiscoveryUnstructured(o *unstructured.Unstructured, in
 
 	return u, nil
 }
+
+func (c *Reconciler) discoveryDataReconcile(ctx context.Context) {
+	discoveryData, err := c.discoverer.DiscoveryData(ctx)
```
Comment (name212) [inline go_lib/cloud-data/reconciler.go:287]: lets add

```
c.logger.Infoln("Start cloud data discovery step")
defer c.logger.Infoln("Finish cloud data discovery step")
```

File: ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go (hunk 5 above)
Comment (name212) [reply ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:106]: hm. Why do we need to request secret in OpenstackDiscoverer?
I think, we will do all works with secret in the `Reconciler`.

Comment (name212) [reply ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:106]: And we do not need to pass kube clienet to discoverer.

File: go_lib/cloud-data/reconciler.go [hunk 9]
```diff
@@ -274,3 +282,106 @@ func (c *Reconciler) cloudDiscoveryUnstructured(o *unstructured.Unstructured, in
...
+	c.cloudRequestErrorMetric.WithLabelValues("discovery_data").Set(0.0)
+
+	for i := 1; i <= 3; i++ {
+		if i > 1 {
+			c.logger.Infoln("Waiting 3 seconds before next attempt")
+			time.Sleep(3 * time.Second)
+		}
+
+		cctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+		data, errGetting := c.k8sClient.Resource(schema.GroupVersionResource{Group: "", Version: "v1", Resource: "secrets"}).Namespace("kube-system").Get(cctx, "d8-cloud-provider-discovery-data", metav1.GetOptions{})
```
Comment (name212) [inline go_lib/cloud-data/reconciler.go:302]: We can use normal (not dynamic) client for secrets. Lets initialize regular go-client and pass to Reconciler

File: go_lib/cloud-data/reconciler.go [hunk 10]
```diff
@@ -274,3 +282,106 @@ func (c *Reconciler) cloudDiscoveryUnstructured(o *unstructured.Unstructured, in
...
+
+		c.updateResourceErrorMetric.WithLabelValues().Set(0.0)
+		return
+	}
+
+	c.updateResourceErrorMetric.WithLabelValues().Set(1.0)
+	c.logger.Errorln("Cannot update cloud data resource. Timed out. See error messages below.")
+}
+
+func (c *Reconciler) discoveryDataUnstructured(o *unstructured.Unstructured, discoveryData []byte) (*unstructured.Unstructured, error) {
```
Comment (name212) [inline go_lib/cloud-data/reconciler.go:352]: with using regular kube-client you do not need to convert to unstructured

File: ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go [hunk 11]
```diff
@@ -84,3 +98,252 @@ func (d *Discoverer) InstanceTypes(_ context.Context) ([]v1alpha1.InstanceType,
...
+		VolumeTypes:              volumeTypes,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal discovery data: %v", err)
+	}
+
+	return discoveryDataJson, nil
+}
+
+func (d *Discoverer) getFlavors(ctx context.Context, provider *gophercloud.ProviderClient) ([]string, error) {
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:174]: We can use InstanceTypes method here

Comment (name212) [reply ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:174]: end empty list is ok

Comment (name212) [reply ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:174]: or use same method inside InstanceTypes method

File: ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go [hunk 12]
```diff
@@ -84,3 +98,252 @@ func (d *Discoverer) InstanceTypes(_ context.Context) ([]v1alpha1.InstanceType,
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to list networks: %v", err)
+	}
+
+	networks, err := networks.ExtractNetworks(allPages)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract networks: %v", err)
+	}
+	if len(networks) == 0 {
+		return nil, errors.New("networks list is empty")
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:265]: empty networks is ok

File: ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/go.mod [hunk 13]
```diff
@@ -22,23 +66,23 @@ require (
...
+	github.com/matttproud/golang_protobuf_extensions v1.0.4 // indirect
 	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
 	github.com/modern-go/reflect2 v1.0.2 // indirect
 	github.com/prometheus/client_model v0.2.0 // indirect
-	github.com/prometheus/common v0.32.1 // indirect
-	github.com/prometheus/procfs v0.7.3 // indirect
+	github.com/prometheus/common v0.37.0 // indirect
+	github.com/prometheus/procfs v0.8.0 // indirect
 	github.com/spf13/pflag v1.0.5 // indirect
-	golang.org/x/net v0.4.0 // indirect
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/go.mod:37]: please use same libraries version with deckhouse

File: modules/030-cloud-provider-aws/images/cloud-data-discoverer/discoverer.go [hunk 14]
```diff
@@ -114,3 +114,7 @@ func (d *Discoverer) InstanceTypes(_ context.Context) ([]v1alpha1.InstanceType,
 
 	return res, nil
 }
+
+func (d *Discoverer) DiscoveryData(ctx context.Context) ([]byte, error) {
+	return []byte("{}"), nil
```
Comment (name212) [inline modules/030-cloud-provider-aws/images/cloud-data-discoverer/discoverer.go:119]: lets return nil and if function return nil without error will not create secret with data.

Comment (name212) [reply modules/030-cloud-provider-aws/images/cloud-data-discoverer/discoverer.go:119]: no secret is better than a knowingly invalid secret

**Review by name212 — CHANGES_REQUESTED** [review]
comments

**Comment by artbin** [comment]
Manual testing:
- `kubectl -n d8-system set image deployment/deckhouse deckhouse=dev-registry.deckhouse.io/sys/deckhouse-oss:pr4793`
```
deployment.apps/deckhouse image updated
```

- `kubectl -n d8-system logs deckhouse-7654fc6dff-frkqb | grep "failed to find secret 'd8-cloud-provider-discovery-data' in namespace 'kube-system'"`

```
{"level":"warning","msg":"failed to find secret 'd8-cloud-provider-discovery-data' in namespace 'kube-system'","time":"2023-06-21T09:14:31Z"}
{"level":"warning","msg":"failed to find secret 'd8-cloud-provider-discovery-data' in namespace 'kube-system'","time":"2023-06-21T09:14:31Z"}
```

- `kubectl -n d8-cloud-provider-openstack logs cloud-data-discoverer-5bdcb6874f-srj5x | grep -A 10 "Start cloud data discoverer"`

```
{"level":"info","msg":"Start cloud data discoverer","time":"2023-06-21T09:14:36Z"}
{"level":"info","msg":"Address:  127.0.0.1:8081","time":"2023-06-21T09:14:36Z"}
{"level":"info","msg":"Checks interval:  1h0m0s","time":"2023-06-21T09:14:36Z"}
{"level":"info","msg":"Start next data discovery","time":"2023-06-21T09:14:36Z"}
{"level":"info","msg":"Start instance type discovery step","time":"2023-06-21T09:14:36Z"}
{"level":"info","msg":"Finish instance type discovery step","time":"2023-06-21T09:14:37Z"}
{"level":"info","msg":"Start cloud data discovery step","time":"2023-06-21T09:14:37Z"}
{"level":"info","msg":"Finish cloud data discovery step","time":"2023-06-21T09:14:37Z"}
{"level":"info","msg":"Finish data discovery","time":"2023-06-21T09:14:37Z"}
```

- `kubectl -n kube-system get secret d8-cloud-provider-discovery-data -o json | jq -r '.data."discovery-data.json"' | base64 -d | jq`

```
{
  "apiVersion": "deckhouse.io/v1alpha1",
  "kind": "OpenStackCloudProviderDiscoveryData",
  "flavors": [
    "m1.medium-50g",
    "m1.xlarge",
    "m1.large-cpu-host-passthrough",
    "m2.large-50g",
    "c8m16d100",
    "m1.large",
    "c8m16d50",
    "m1.medium",
    "m1.large-cpu-host-model",
    "m1.tiny",
    "m1.large-50g",
    "m1.xsmall",
    "m1.small"
  ],
  "additionalNetworks": [
    "public",
    "dev",
    "shared"
  ],
  "additionalSecurityGroups": [
    "default",
    "dev",
    "dev-frontend"
  ],
  "defaultImageName": "ubuntu-22-04-cloud-amd64",
  "images": [
    "centos-9-x86_64-cloud",
    "alse-vanilla-1.7.3-cloud-adv-mg9.1.2",
    "alse-vanilla-1.7.3-cloud-max-mg9.1.2",
    "almalinux-8.7.x86_64",
    "almalinux-9.1.x86_64",
    "redos-standard-7.3.2",
    "alt-p10-cloud-amd64",
    "fedora-coreos-36.20221014.3.1-openstack.x86_64",
    "alse-vanilla-1.7.2-cloud-mg7.2.0",
    "ubuntu-22-04-cloud-amd64",
    "debian-11-cloud-amd64",
    "orel-vanilla-2.12.43-cloud-mg6.5.0",
    "centos-8-cloud-amd64",
    "deepin-desktop-community-20.4-amd64-iso",
    "alteros-server-cloud",
    "alse-vanilla-1.7.1-cloud-mg6.4.0",
    "debian-9-cloud-amd64",
    "orel-vanilla-2.12.43-cloud-mg6",
    "debian-10-cloud-amd64",
    "orel-vanilla-2.12.43-cloud",
    "ubuntu-20-04-cloud-amd64",
    "centos-7-x86_64-cloud-2003",
    "ubuntu-18-04-cloud-amd64-kubernetes-1.15.3",
    "ubuntu-16-04-cloud-amd64",
    "ubuntu-18-04-cloud-amd64"
  ],
  "mainNetwork": "dev",
  "zones": [
    "nova"
  ],
  "volumeTypes": [
    {
      "id": "cc728a7f-787d-4a3d-ae3a-bf7611cda23e",
      "name": "__DEFAULT__",
      "description": "Default Volume Type",
      "isPublic": true
    },
    {
      "id": "b5637549-2e1a-4cf5-adb9-d854f52a9865",
      "name": "ceph-ssd",
      "isPublic": true
    }
  ]
}
```

- `kubectl -n d8-system exec -it deploy/deckhouse -- deckhouse-controller module values cloud-provider-openstack 2>/dev/null | yq .cloudProviderOpenstack.internal.discoveryData`

```
additionalNetworks:
  - public
  - dev
  - shared
additionalSecurityGroups:
  - default
  - dev
  - dev-frontend
apiVersion: deckhouse.io/v1alpha1
defaultImageName: ubuntu-22-04-cloud-amd64
flavors:
  - m1.medium-50g
  - m1.xlarge
  - m1.large-cpu-host-passthrough
  - m2.large-50g
  - c8m16d100
  - m1.large
  - c8m16d50
  - m1.medium
  - m1.large-cpu-host-model
  - m1.tiny
  - m1.large-50g
  - m1.xsmall
  - m1.small
images:
  - centos-9-x86_64-cloud
  - alse-vanilla-1.7.3-cloud-adv-mg9.1.2
  - alse-vanilla-1.7.3-cloud-max-mg9.1.2
  - almalinux-8.7.x86_64
  - almalinux-9.1.x86_64
  - redos-standard-7.3.2
  - alt-p10-cloud-amd64
  - fedora-coreos-36.20221014.3.1-openstack.x86_64
  - alse-vanilla-1.7.2-cloud-mg7.2.0
  - ubuntu-22-04-cloud-amd64
  - debian-11-cloud-amd64
  - orel-vanilla-2.12.43-cloud-mg6.5.0
  - centos-8-cloud-amd64
  - deepin-desktop-community-20.4-amd64-iso
  - alteros-server-cloud
  - alse-vanilla-1.7.1-cloud-mg6.4.0
  - debian-9-cloud-amd64
  - orel-vanilla-2.12.43-cloud-mg6
  - debian-10-cloud-amd64
  - orel-vanilla-2.12.43-cloud
  - ubuntu-20-04-cloud-amd64
  - centos-7-x86_64-cloud-2003
  - ubuntu-18-04-cloud-amd64-kubernetes-1.15.3
  - ubuntu-16-04-cloud-amd64
  - ubuntu-18-04-cloud-amd64
kind: OpenStackCloudProviderDiscoveryData
mainNetwork: dev
volumeTypes:
  - description: Default Volume Type
    id: cc728a7f-787d-4a3d-ae3a-bf7611cda23e
    isPublic: true
    name: __DEFAULT__
  - id: b5637549-2e1a-4cf5-adb9-d854f52a9865
    isPublic: true
    name: ceph-ssd
zones:
  - nova
```

File: ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go [hunk 15]
```diff
@@ -7,15 +7,21 @@ package main
...
 	"github.com/gophercloud/gophercloud"
 	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/v3/volumetypes"
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/flavors"
+	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/images"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/security/groups"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
 	log "github.com/sirupsen/logrus"
 	"k8s.io/apimachinery/pkg/api/resource"
+	"os"
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/images/cloud-data-discoverer/discoverer.go:22]: Imports not sorted

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @artbin. I've started the workflow for label 'e2e/run/openstack'...

File: ee/modules/030-cloud-provider-openstack/hooks/discover_test.go [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+  },
+  {
+	"name": "ssd-r1",
+	"type": "SSD R1"
+  }
+]
+`))
+			Expect(c.ValuesGet("cloudProviderOpenstack.internal.defaultStorageClass").String()).To(Equal(`other-bar`))
+		})
+	})
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/hooks/discover_test.go:258]: Need to add test with no discovery data secret and:
   no storage classes - hook should not fail
   have storage classes - storage classes should in  values

**Review by name212 — CHANGES_REQUESTED** [review]
2 tests

File: ee/modules/030-cloud-provider-openstack/hooks/discover.go [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+			NameSelector: &types.NameSelector{
+				MatchNames: []string{"d8-cloud-provider-discovery-data"},
+			},
+			FilterFunc: applyCloudProviderDiscoveryDataSecretFilter,
+		},
+		{
+			Name:       "storage_classes",
+			ApiVersion: "storage.k8s.io/v1",
+			Kind:       "StorageClass",
+			FilterFunc: applyStorageClassFilter,
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/hooks/discover.go:49]: I forget about one thing. Cluster admin can add own storage class. We need to filter storage class by label selector `heritage: deckhouse` as [here](https://github.com/deckhouse/deckhouse/blob/468ae94c1d944d253b2d62a5e26050d1efa0a823/modules/340-monitoring-kubernetes/hooks/storage_class_cloud_manual.go#L61)

File: ee/modules/030-cloud-provider-openstack/hooks/discover.go [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+	"unicode"
+
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
+	log "github.com/sirupsen/logrus"
```
Comment (name212) [inline ee/modules/030-cloud-provider-openstack/hooks/discover.go:18]: do not use `logrus` in hooks. Please use `input.LogEntry` as [here](https://github.com/deckhouse/deckhouse/blob/468ae94c1d944d253b2d62a5e26050d1efa0a823/modules/040-control-plane-manager/hooks/etcd_quota_backend_bytes.go#L240C4-L240C18)

File: ee/modules/030-cloud-provider-openstack/hooks/discover.go (hunk 17 above)
Comment (name212) [reply ee/modules/030-cloud-provider-openstack/hooks/discover.go:49]: and add test for cover this case

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @artbin. I've started the workflow for label 'e2e/run/openstack'...

**Review by name212 — APPROVED** [review]

**Review by nabokihms — APPROVED** [review]

**Review by konstantin-axenov — APPROVED** [review]

**Pull request merged**